include/api/CDataFrameAnalysisSpecification.h

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+#ifndef INCLUDED_ml_api_CDataFrameAnalysisSpecification_h
+#define INCLUDED_ml_api_CDataFrameAnalysisSpecification_h
+
+#include <core/CFastMutex.h>
+
+#include <api/ImportExport.h>
+
+#include <rapidjson/fwd.h>
+
+#include <cstddef>
+#include <memory>
+#include <string>
+#include <thread>
+#include <vector>
+
+namespace ml {
+namespace core {
+class CDataFrame;
+}
+namespace api {
+class CDataFrameAnalysisRunner;
+class CDataFrameAnalysisRunnerFactory;
+
+//! \brief Parses a complete specification for running a core::CDataFrame analysis
+//! and supports launching that analysis on a specified frame object.
+//!
+//! DESCRIPTION:\n
+//! This manages extracting all configuration for a particular analysis from a JSON
+//! header which is passed to the data_frame_analyzer command before any data. This
+//! creates and owns an analysis runner object which is also configured by the header.
+//! The analysis is run asynchronously via the CDataFrameAnalysisSpecification::run
+//! method which returns a handle to the runner to retrieve progress and errors.
+class API_EXPORT CDataFrameAnalysisSpecification {
+public:
+    using TStrVec = std::vector<std::string>;
+    using TRunnerPtr = std::unique_ptr<CDataFrameAnalysisRunner>;
+    using TRunnerFactoryPtr = std::unique_ptr<CDataFrameAnalysisRunnerFactory>;
+    using TRunnerFactoryPtrVec = std::vector<TRunnerFactoryPtr>;
+
+public:
+    //! Inititialize from a JSON object.
+    //!
+    //! The specification has the following expected form:
+    //! <CODE>
+    //! {
+    //!   "rows": <integer>,
+    //!   "cols": <integer>,
+    //!   "memory_limit": <integer>,
+    //!   "threads": <integer>,
+    //!   "analysis": {
+    //!     "name": <string>,
+    //!     "parameters": <object>
+    //!   }
+    //! }
```

Comment:
We might want to rename some of these fields to match what will be used in the Java side configuration objects.  That will reduce confusion in the future.  But as far as this PR is concerned it can be merged with temporary names and they can be renamed in a future PR.

/cc @dimitris-athanasiou

Reply:
I'll leave a TODO in the code. This is really a placeholder and we should definitely standardise naming.

Reply:
[0839984](https://github.com/elastic/ml-cpp/pull/306/commits/0839984d906801c86b4079f03d89ad4d39ab15de)